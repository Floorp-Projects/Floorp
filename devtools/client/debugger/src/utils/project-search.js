/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// Maybe reuse file search's functions?

import React from "react";

export function highlightMatches(lineMatch) {
  const { value, matchIndex, match } = lineMatch;
  const len = match.length;

  return (
    <span className="line-value">
      <span className="line-match" key={0}>
        {value.slice(0, matchIndex)}
      </span>
      <span className="query-match" key={1}>
        {value.substr(matchIndex, len)}
      </span>
      <span className="line-match" key={2}>
        {value.slice(matchIndex + len, value.length)}
      </span>
    </span>
  );
}
