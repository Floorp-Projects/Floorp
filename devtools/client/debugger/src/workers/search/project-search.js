/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// Maybe reuse file search's functions?

import getMatches from "./get-matches";

export function findSourceMatches(sourceId, content, queryText) {
  if (queryText == "") {
    return [];
  }

  const modifiers = {
    caseSensitive: false,
    regexMatch: false,
    wholeWord: false,
  };

  const text = content.value;
  const lines = text.split("\n");

  return getMatches(queryText, text, modifiers).map(({ line, ch }) => {
    const { value, matchIndex } = truncateLine(lines[line], ch);
    return {
      sourceId,
      line: line + 1,
      column: ch,
      matchIndex,
      match: queryText,
      value,
    };
  });
}

// This is used to find start of a word, so that cropped string look nice
const startRegex = /([ !@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?])/g;
// Similarly, find
const endRegex = new RegExp(
  [
    "([ !@#$%^&*()_+-=[]{};':\"\\|,.<>/?])",
    '[^ !@#$%^&*()_+-=[]{};\':"\\|,.<>/?]*$"/',
  ].join("")
);

function truncateLine(text, column) {
  if (text.length < 100) {
    return {
      matchIndex: column,
      value: text,
    };
  }

  // Initially take 40 chars left to the match
  const offset = Math.max(column - 40, 0);
  // 400 characters should be enough to figure out the context of the match
  const truncStr = text.slice(offset, column + 400);
  let start = truncStr.search(startRegex);
  let end = truncStr.search(endRegex);

  if (start > column) {
    // No word separator found before the match, so we take all characters
    // before the match
    start = -1;
  }
  if (end < column) {
    end = truncStr.length;
  }
  const value = truncStr.slice(start + 1, end);

  return {
    matchIndex: column - start - offset - 1,
    value,
  };
}
