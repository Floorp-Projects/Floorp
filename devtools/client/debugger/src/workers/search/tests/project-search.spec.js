/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import { findSourceMatches } from "../project-search";

const text = `
  function foo() {
    foo();
  }
`;

describe("project search", () => {
  const emptyResults = [];

  it("finds matches", () => {
    const needle = "foo";
    const content = {
      type: "text",
      value: text,
      contentType: undefined,
    };

    const matches = findSourceMatches("bar.js", content, needle);
    expect(matches).toMatchSnapshot();
  });

  it("finds no matches in source", () => {
    const needle = "test";
    const content = {
      type: "text",
      value: text,
      contentType: undefined,
    };
    const matches = findSourceMatches("bar.js", content, needle);
    expect(matches).toEqual(emptyResults);
  });
});
