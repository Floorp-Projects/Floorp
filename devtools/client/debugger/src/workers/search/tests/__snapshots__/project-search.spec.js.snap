// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project search finds matches 1`] = `
Array [
  Object {
    "column": 11,
    "line": 2,
    "match": "foo",
    "matchIndex": 11,
    "sourceId": "bar.js",
    "value": "  function foo() {",
  },
  Object {
    "column": 4,
    "line": 3,
    "match": "foo",
    "matchIndex": 4,
    "sourceId": "bar.js",
    "value": "    foo();",
  },
]
`;
